Exclude file name from detect common prefix

mecampbellsoup opened the pull request:
## Current Behavior

Due to `detect_common_prefix` implementation, it is possible to return a 'common prefix' that includes shared portions of the actual file names contained in the files list.

However, looking at the code in the `build` method and specifically at how [the prefix is assumed to be a path](https://github.com/cargoserver/ngzip/blob/master/lib/ngzip/builder.rb#L25), it seems as if the intention is to automatically include any shared _leading path_ prefixes without looking into the file names.

I also changed the `encode` method to use `ERB::Util.url_encode` as that is more idiomatic these days in Ruby; and made some small noop formatting changes.

dup2 commented:
Thanks for this PR!
It is correct that only _leading path_ prefixes should be used for this, not file names.

dup2 commented:
When running the tests, I get a lot of errors like `NameError: uninitialized constant Ngzip::Builder::ERB` for this, maybe you are using this in a Rails context where ERB is already loaded.

mecampbellsoup commented:
> When running the tests, I get a lot of errors like `NameError: uninitialized constant Ngzip::Builder::ERB` for this, maybe you are using this in a Rails context where ERB is already loaded.

Ah, apologies, just need to `require 'erb'` - part of the core library in Ruby. https://ruby-doc.org/stdlib-2.7.1/libdoc/erb/rdoc/ERB.html

mecampbellsoup commented:
@dup2 BTW, how can I run the tests? Once you tell me I will add to the README in my PR as well.

mecampbellsoup commented:
> @dup2 BTW, how can I run the tests? Once you tell me I will add to the README in my PR as well.

```
➜  ngzip git:(mc/exclude-file-name-from-detect-common-prefix) ✗ rake test
/Users/mattysoup/.rbenv/versions/2.6.5/bin/ruby -w -I"lib:lib" -I"/Users/mattysoup/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-13.0.1/lib" "/Users/mattysoup/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-13.0.1/lib/rake/rake_test_loader.rb" "test/lib/ngzip/builder_test.rb" "test/lib/ngzip/version_test.rb"

File does not exist: minitest/autorun

rake aborted!
Command failed with status (1): [ruby -w -I"lib:lib" -I"/Users/mattysoup/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-13.0.1/lib" "/Users/mattysoup/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-13.0.1/lib/rake/rake_test_loader.rb" "test/lib/ngzip/builder_test.rb" "test/lib/ngzip/version_test.rb" ]
/Users/mattysoup/.rbenv/versions/2.6.5/bin/bundle:23:in `load'
/Users/mattysoup/.rbenv/versions/2.6.5/bin/bundle:23:in `<main>'
Tasks: TOP => test
(See full trace by running task with --trace)
```

mecampbellsoup commented:
@dup2 I got the tests running and fixed up:

```
(⎈ tenant-matt-user-tenant-matt-tenant-cluster.chi.atlanticcrypto.com:tenant-matt)➜  ngzip git:(master) rake test
/Users/mattysoup/.rbenv/versions/2.6.5/bin/ruby -w -I"lib:lib" -I"/Users/mattysoup/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-13.0.1/lib" "/Users/mattysoup/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-13.0.1/lib/rake/rake_test_loader.rb" "test/lib/ngzip/builder_test.rb" "test/lib/ngzip/version_test.rb"
Started with run options --seed 60842

DEPRECATED: global use of must_equal from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:48. Use _(obj).must_equal instead. This will fail in Minitest 6.--=---=---=---=---=---] 0% Time: 00:00:00,  ETA: ??:??:??
DEPRECATED: global use of wont_be_nil from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:32. Use _(obj).wont_be_nil instead. This will fail in Minitest 6.
DEPRECATED: global use of wont_be_nil from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:36. Use _(obj).wont_be_nil instead. This will fail in Minitest 6.
DEPRECATED: global use of must_equal from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:40. Use _(obj).must_equal instead. This will fail in Minitest 6.
DEPRECATED: global use of wont_be_nil from /Users/mattysoup/code/ngzip/test/lib/ngzip/version_test.rb:6. Use _(obj).wont_be_nil instead. This will fail in Minitest 6.
DEPRECATED: global use of must_equal from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:101. Use _(obj).must_equal instead. This will fail in Minitest 6.
DEPRECATED: global use of must_equal from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:143. Use _(obj).must_equal instead. This will fail in Minitest 6.
DEPRECATED: global use of must_equal from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:136. Use _(obj).must_equal instead. This will fail in Minitest 6.
DEPRECATED: global use of must_equal from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:76. Use _(obj).must_equal instead. This will fail in Minitest 6.
DEPRECATED: global use of must_equal from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:129. Use _(obj).must_equal instead. This will fail in Minitest 6.
DEPRECATED: global use of must_equal from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:107. Use _(obj).must_equal instead. This will fail in Minitest 6.
DEPRECATED: global use of must_equal from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:86. Use _(obj).must_equal instead. This will fail in Minitest 6.
DEPRECATED: global use of must_equal from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:57. Use _(obj).must_equal instead. This will fail in Minitest 6.
DEPRECATED: global use of must_equal from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:95. Use _(obj).must_equal instead. This will fail in Minitest 6.
DEPRECATED: global use of must_equal from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:67. Use _(obj).must_equal instead. This will fail in Minitest 6.
DEPRECATED: global use of must_equal from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:62. Use _(obj).must_equal instead. This will fail in Minitest 6.
DEPRECATED: global use of must_equal from /Users/mattysoup/code/ngzip/test/lib/ngzip/builder_test.rb:149. Use _(obj).must_equal instead. This will fail in Minitest 6.
  19/19: [===============================================================================================================================================================================] 100% Time: 00:00:00, Time: 00:00:00

Finished in 0.00674s
19 tests, 19 assertions, 0 failures, 0 errors, 0 skips
```

Please note I added another commit fixing a situation where the in-manifest file names could potentially include leading slashes thus breaking Windows File Explorer extractions of zipfiles, e.g. `builder.build('/path/to/directory/on/nginx/server/')` which before my change would generate a manifest like: 

```
- 466 /path/to/directory/on/nginx/server/0001.png /001.png
```
(The 3rd value would be URL encoded, just omitting that for this example.)

See [`7eb57a4` (#5)](https://github.com/cargoserver/ngzip/pull/5/commits/7eb57a4dd1210e5c598cda908bbcf609871b47dd).

dup2 commented on test/lib/ngzip/builder_test.rb:
```diff
@@ -5,23 +5,27 @@
 
   it 'must support the static method :build' do
-    Ngzip.respond_to?(:build).must_equal true
+    expect(Ngzip.respond_to?(:build)).must_equal true
```
Isn't this kind of double testing? I think `(foo).must_equal true` is enough for testing the boolean result here which is returned by `respond_to?`

mecampbellsoup replied:
Oh, this change is to address a warning that MiniTest generates: https://stackoverflow.com/a/58128950/2517033

dup2 replied:
Ah, I see - did only know the "_()" version before. As I use rspec matchers also with minitest this seemed like a duplication.

mecampbellsoup replied:
@dup2 LMK if anything else on this you want changed!

dup2 commented:
Thanks for all the updates!
